test/core/call/request_buffer_test.cc

```diff
@@ -0,0 +1,573 @@
+// Copyright 2024 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "src/core/call/request_buffer.h"
+
+#include "gtest/gtest.h"
+
+#include "test/core/promise/poll_matcher.h"
+
+using testing::Mock;
+using testing::StrictMock;
+
+namespace grpc_core {
+
+namespace {
+void CrashOnParseError(absl::string_view error, const Slice& data) {
+  LOG(FATAL) << "Failed to parse " << error << " from "
+             << data.as_string_view();
+}
+
+// A mock activity that can be activated and deactivated.
+class MockActivity : public Activity, public Wakeable {
+ public:
+  MOCK_METHOD(void, WakeupRequested, ());
+
+  void ForceImmediateRepoll(WakeupMask /*mask*/) override { WakeupRequested(); }
+  void Orphan() override {}
+  Waker MakeOwningWaker() override { return Waker(this, 0); }
+  Waker MakeNonOwningWaker() override { return Waker(this, 0); }
+  void Wakeup(WakeupMask /*mask*/) override { WakeupRequested(); }
+  void WakeupAsync(WakeupMask /*mask*/) override { WakeupRequested(); }
+  void Drop(WakeupMask /*mask*/) override {}
+  std::string DebugTag() const override { return "MockActivity"; }
+  std::string ActivityDebugTag(WakeupMask /*mask*/) const override {
+    return DebugTag();
+  }
+
+  void Activate() {
+    if (scoped_activity_ == nullptr) {
+      scoped_activity_ = std::make_unique<ScopedActivity>(this);
+    }
+  }
+
+  void Deactivate() { scoped_activity_.reset(); }
+
+ private:
+  std::unique_ptr<ScopedActivity> scoped_activity_;
+};
+
+#define EXPECT_WAKEUP(activity, statement)                                 \
+  EXPECT_CALL((activity), WakeupRequested()).Times(::testing::AtLeast(1)); \
+  statement;                                                               \
+  Mock::VerifyAndClearExpectations(&(activity));
+
+ClientMetadataHandle TestMetadata() {
+  ClientMetadataHandle md = Arena::MakePooledForOverwrite<ClientMetadata>();
+  md->Append("key", Slice::FromStaticString("value"), CrashOnParseError);
+  return md;
+}
+
+MessageHandle TestMessage(int index = 0) {
+  return Arena::MakePooled<Message>(
+      SliceBuffer(Slice::FromCopiedString(absl::StrCat("message ", index))), 0);
+}
+
+MATCHER(IsTestMetadata, "") {
+  if (arg == nullptr) return false;
+  std::string backing;
+  if (arg->GetStringValue("key", &backing) != "value") {
```

Comment:
I think we'll get a better error message when the test fails if we write it like this:

```
return ::testing::ExplainMatchResult("value", arg->GetStringValue("key", &backing), result_listener);
```

Reply:
before:

```
test/core/call/request_buffer_test.cc:538: Failure
Value of: *value1
Expected: is test metadata
  Actual: (ptr = 0x55c331da96c0, value = 568-byte object <00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 ... 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-99 DA-31 C3-55 00-00 40-99 DA-31 C3-55 00-00 40-99 DA-31 C3-55 00-00>), key: 5 bytes redacted by allow listing.
```

after:

```
test/core/call/request_buffer_test.cc:535: Failure
Value of: *value1
Expected: is test metadata
  Actual: (ptr = 0x55d04be316c0, value = 568-byte object <00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 ... 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 C0-1B E3-4B D0-55 00-00 00-1C E3-4B D0-55 00-00 00-1C E3-4B D0-55 00-00>)
```

Reply:
(i'm going to skip these suggestions)

Reply:
Interesting!  I was under the impression that `ExplainMatchResult()` would show clearer data, but I guess that doesn't work the way I thought it did.